### PR TITLE
[Feature] CAS Institution Login [#CAS-3]

### PR DESCRIPTION
**Pull request by cslzchen**
## Tickets and related PRs

Ticket: https://openscience.atlassian.net/browse/CAS-3
OSF-PR: https://github.com/CenterForOpenScience/osf.io/pull/6082
## Description

Use CAS for institution login.
## Deployment Considerations

```
cas.osf.login.url=/login?
cas.institution.login.url=/login?campaign=institution
```
### UI
- Slightly-modified CAS Login Page

cas-login-page
- New CAS Institution Login Page

cas-institution-login-page
- Old OSF Institution Login Page

osf-inst-login-old
### CAS Backend
- Refactor `OpenScienceFrameworkLogoutHandler` to `OpenScienceFrameworkInstitutionHandler`:
  add necessary institution fields;
  add `getInstitutionLoginUrls` and return a map of institutions' name and login url.
- **Action State**: `beforeLogin`
  **Action Event**: `OpenScienceFrameworkLoginHandler.beforeLogin`
  A decision state that chooses between normal OSF login and institution login, which is also scalable for campaign login.
- **Action State**: `institutionLogin`
  **Action Event**: `OpenScienceFrameworkInstitutionLoginHandler.getInstitutions`
  Retrieve institutions from DB and put a map in flow.
-  **View State**: `viewInstitutionLoginForms`
  Target alternative login page `casInstitutionLoginView.jsp`
- Modified CAS login flow:

```
    -->> remoteAuthenticate
        -->> success: sendTGT
        -->> failure: authFailure
        -->> error: beforeLogin

   -->> beforeLogin
        -->> osf: viewLoginForm
        -->> institution: institutionLogin

    -->> institutionLogin
        -->> success: viewInstitutionLoginForms
```
### OSF Backend

Please refer to [OSF-PR#6082](https://github.com/CenterForOpenScience/osf.io/pull/6082) for more information.
## Side Effects

No

### CODE REVIEW

File: cas-server-support-osf/src/main/java/io/cos/cas/adaptors/mongodb/OpenScienceFrameworkLogoutHandler.java [hunk 1]
```diff
@@ -1,118 +0,0 @@
-/*
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/io/cos/cas/adaptors/mongodb/OpenScienceFrameworkLogoutHandler.java:1]: TO CR: this file is renamed to `OpenScienceFrameworkInstitutionAuthenticationHandler`. The one commit for the rename consider it `mv` but the PR does not. There are a two main changes: 1. add all institution related fields to `OpenScienceFrameworkInstitution` 2. add a method to return a map of institution name and login url.

File: cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
+/*
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java]: To CR: for the new `action-state: institutionLogin`, triggered by `Event(institution)` in `action-state: remoteAuthenticate` and go on to the new `view-state: viewInstitutionLoginForm`.

File: cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkTerminateSessionAction.java [hunk 3]
```diff
@@ -130,9 +130,9 @@ public Event terminate(final RequestContext context) {
 
         // if users logged in through their institutions, redirect to institution logout endpoint
         if (remotePrincipal && institutionId != null) {
-            final String institutionLogoutUrl = this.logoutHandler.findInstitutionLogoutUrlById(institutionId);
+            final String institutionLogoutUrl = this.institutionHandler.findInstitutionLogoutUrlById(institutionId);
             if (institutionLogoutUrl == null) {
-                logger.warn("Institution {} does not have logout url, use default logout redirection instead", institutionId);
+                logger.warn("Institution {} does not have an dedicated logout url, use default logout redirection instead", institutionId);
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkTerminateSessionAction.java]: To CR: I will fix this typo, should be `a dedicated`.

File: etc/cas.properties [hunk 4]
```diff
@@ -32,10 +32,15 @@ google.analytics.id=
 google.analytics.autoLink=
 
 ##
+# CAS Login URLs
+#
+cas.osf.login.url=/login?
+cas.institution.login.url=/login?campaign=institution
+
```
Comment (cslzchen) [inline etc/cas.properties:39]: For CR: need to update this in deployment.

File: cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java [hunk 5]
```diff
@@ -167,12 +167,23 @@ public String getInstitutionId() {
...
 
+    /**
+     * Check institution login. Return true if `campaign=institution` is present in request parameters.
+     *
+     * @param context The request context
+     * @return Boolean
+     */
+    protected boolean isInstitution(final RequestContext context) {
+        final String campaign = context.getRequestParameters().get("campaign");
+        return campaign != null && "institution".equals(campaign);
```
Comment (icereval) [inline cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java]: may not need `campaign != null &&`

File: cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                throw new AssertionError("UTF-8 is unknown");
+            }
+        }
+        context.getFlowScope().put("target", target);
+        context.getFlowScope().put("campaign", "INSTITUTION");
+        return new Event(this, "success");
+    }
+
+    /**
+     * Sort a Map by value. Return teh sorted Map.
```
Comment (icereval) [inline cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java]: typo `teh`

File: cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+     * @param context The request context
+     * @return Event
+     */
+    public Event getInstitutions(final RequestContext context) {
+        final Map<String, String> institutions = this.institutionHandler.getInstitutionLogin();
+        institutions.put("", " -- select an institution -- ");
+        final Map<String, String> sortedInstitutions = sortByValue(institutions);
+        logger.info(String.format("Institutions loaded: %s", sortedInstitutions.toString()));
+        context.getFlowScope().put("institutions", sortedInstitutions);
+        String target = "";
```
Comment (icereval) [inline cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java]: should target default to encoded `/login`?

File: cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    /**
+     * Sort a Map by value. Return teh sorted Map.
+     * @param map The Map to be sorted
+     * @param <K> The Type of the Key
+     * @param <V> The Type of the Value
+     * @return Map&lt;K,V&gt;
+     */
+    private static <K, V extends Comparable<? super V>> Map<K, V> sortByValue(final Map<K, V> map) {
```
Comment (icereval) [inline cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java]: is this utility method from somewhere we can source?

File: cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkTerminateSessionAction.java [hunk 9]
```diff
@@ -63,32 +63,32 @@
     @NotNull
     private final CookieRetrievingCookieGenerator warnCookieGenerator;
 
-    /** Logout handler. */
+    /** Institution authentication handler for logout. */
```
Comment (icereval) [inline cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkTerminateSessionAction.java:66]: Institution handler

File: cas-server-support-osf/src/main/java/io/cos/cas/adaptors/mongodb/OpenScienceFrameworkInstitutionAuthenticationHandler.java [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The Open Science Framework Institution Authentication Handler.
+ *
+ * @author Longze Chen
+ * @since 4.1.0
+ */
+public class OpenScienceFrameworkInstitutionAuthenticationHandler {
```
Comment (icereval) [inline cas-server-support-osf/src/main/java/io/cos/cas/adaptors/mongodb/OpenScienceFrameworkInstitutionAuthenticationHandler.java]: OpenScienceFrameworkInstitutionHandler

File: cas-server-webapp/src/main/webapp/WEB-INF/spring-configuration/institutionAuthenticationManager.xml [hunk 11]
```diff
@@ -22,8 +22,12 @@
 <beans xmlns="http://www.springframework.org/schema/beans"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xmlns:p="http://www.springframework.org/schema/p"
+       xmlns:c="http://www.springframework.org/schema/c"
        xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-3.2.xsd">
 
-    <bean id="openScienceFrameworkLogoutHandler" class="io.cos.cas.adaptors.mongodb.OpenScienceFrameworkLogoutHandler"
+    <bean id="openScienceFrameworkInstitutionAuthenticationHandler" class="io.cos.cas.adaptors.mongodb.OpenScienceFrameworkInstitutionAuthenticationHandler"
```
Comment (icereval) [inline cas-server-webapp/src/main/webapp/WEB-INF/spring-configuration/institutionAuthenticationManager.xml]: openScienceFrameworkInstitutionHandler

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/casInstitutionLoginView.jsp [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    <section class="row">
+        <div class="btn-submit">
+            <input type="button" name="submit" value="<spring:message code="screen.welcome.button.login" />" onclick="institutionLogin()">
+            <input type="hidden" id="target" value="${target}" />
+        </div>
+    </section>
+    <br>
+    <script>
+        function institutionLogin () {
+            var form = document.getElementById("institution-form-select");
```
Comment (icereval) [inline cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/casInstitutionLoginView.jsp]: trivial, use single quotes in JS when possible.

https://github.com/airbnb/javascript/blob/master/README.md#strings--quotes

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/casInstitutionLoginView.jsp [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    <br>
+    <section class="row">
+        <div class="btn-submit">
+            <input type="button" name="submit" value="<spring:message code="screen.welcome.button.login" />" onclick="institutionLogin()">
+            <input type="hidden" id="target" value="${target}" />
+        </div>
+    </section>
+    <br>
+    <script>
+        function institutionLogin () {
```
Comment (icereval) [inline cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/casInstitutionLoginView.jsp:47]: move JS to backend `getInstitutions` method

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/bottom.jsp [hunk 14]
```diff
@@ -21,16 +21,25 @@
...
     <div class="row" style="text-align: center;">
-        <hr style="margin-bottom: 1em;">
-        <spring:eval var="osfInstitutionLoginUrl" expression="@casProperties.getProperty('osf.institutionLogin.url')" />
-        <a id="institution-login" href="${osfInstitutionLoginUrl}${not empty param.service ? '&redirect_url=' : ''}${fn:escapeXml(param.service)}">Login&nbsp;Through&nbsp;Your&nbsp;Institution</a>&nbsp;&nbsp;
+        <hr>
+        <br>
+        <c:choose>
+            <c:when test="${campaign == 'INSTITUTION'}">
+                <spring:eval var="osfLoginUrl" expression="@casProperties.getProperty('cas.osf.login.url')" />
+                <a id="alternative-osf" href="${osfLoginUrl}${not empty param.service ? 'service=' : ''}${fn:escapeXml(param.service)}">Back to Non-institution Login</a>&nbsp;&nbsp;
```
Comment (icereval) [inline cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/bottom.jsp]: <strike>Back to </strike>

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/bottom.jsp [hunk 15]
```diff
@@ -21,16 +21,25 @@
...
+            </c:when>
+            <c:otherwise>
+                <spring:eval var="institutionLoginUrl" expression="@casProperties.getProperty('cas.institution.login.url')" />
+                <a id="aleternative-institution" href="${institutionLoginUrl}${not empty param.service ? '&service=' : ''}${fn:escapeXml(param.service)}">Login through Your Institution</a>&nbsp;&nbsp;
```
Comment (icereval) [inline cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/bottom.jsp:33]: change to flow, no longer going over to OSF page

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/top.jsp [hunk 16]
```diff
@@ -28,32 +28,47 @@
...
+                        <span class="title-full">Open&nbsp;Science&nbsp;Framework</span>
+                        <span class="title-abbr">Open&nbsp;Science<br>Framework</span>
+                        <c:if test="${campaign eq 'INSTITUTION'}"><span> for&nbsp;Institutions</span></c:if>
+                    </span>
+                </div>
+                <div class="responsive">
+                    <c:if test="${campaign eq 'INSTITUTION'}">
+                        <div id="description">
+                            <br><br><spring:message code="screen.institution.login.message.a" />
+                            <spring:message code="screen.institution.login.message.b" />
```
Comment (icereval) [inline cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/top.jsp]: `.a/b`?

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/top.jsp [hunk 17]
```diff
@@ -28,32 +28,47 @@
...
+                <div class="center">
+                    <spring:eval var="osfUrl" expression="@casProperties.getProperty('osf.url')" />
+                    <a id="logo" class="center" href="${osfUrl}" title="<spring:message code="logo.title" />">Open Science Framework | Sign In</a>
+                </div>
+                <br>
+                <div class="center">
+                    <span id="title">
+                        <span class="title-full">Open&nbsp;Science&nbsp;Framework</span>
+                        <span class="title-abbr">Open&nbsp;Science<br>Framework</span>
+                        <c:if test="${campaign eq 'INSTITUTION'}"><span> for&nbsp;Institutions</span></c:if>
```
Comment (icereval) [inline cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/top.jsp]: use `OSF Institutions`

Comment (icereval) [reply cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/top.jsp]: should be able to remove resize exception

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/top.jsp [hunk 18]
```diff
@@ -28,32 +28,47 @@
...
+                    </span>
+                </div>
+                <div class="responsive">
+                    <c:if test="${campaign eq 'INSTITUTION'}">
```
Comment (icereval) [inline cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/top.jsp:71]: In general campaign should be a data structure passed to the front end, eventually surfaced from the database.

File: cas-server-webapp/src/main/webapp/css/cas.css [hunk 19]
```diff
@@ -249,18 +282,15 @@ footer a:link, footer a:visited {
   #sidebar .sidebar-content { padding: 0; }
 }
 
-
-#fm1 .row .btn-oauth,
-#fm1 .row .btn-oauth:hover,
-#fm1 .row .btn-oauth:active {
+#fm1 .row .btn-oauth {
```
Comment (icereval) [inline cas-server-webapp/src/main/webapp/css/cas.css:285]: test in safari,ie,firefox,chrome

File: etc/cas.properties (hunk 4 above)
Comment (icereval) [reply etc/cas.properties:39]: if not already done, please note in the pr description

Comment (icereval) [reply etc/cas.properties:39]: e.g. https://github.com/CenterForOpenScience/modular-file-renderer/pull/223

**Review by icereval — COMMENTED** [review]
first pass

File: cas-server-webapp/src/main/webapp/WEB-INF/webflow/login/login-webflow.xml [hunk 20]
```diff
@@ -97,8 +97,17 @@
         <transition on="success" to="sendTicketGrantingTicket" />
         <transition on="authenticationFailure" to="handleAuthenticationFailure" />
         <transition on="error" to="viewLoginForm" />
+        <transition on="institution" to="institutionLogin" />
```
Comment (icereval) [inline cas-server-webapp/src/main/webapp/WEB-INF/webflow/login/login-webflow.xml:105]: adjust `error` state to send to a campaign decision step.

File: cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java (hunk 8 above)
Comment (cslzchen) [reply cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java]: No reliable package found.

File: cas-server-webapp/src/main/webapp/WEB-INF/webflow/login/login-webflow.xml (hunk 20 above)
Comment (cslzchen) [reply cas-server-webapp/src/main/webapp/WEB-INF/webflow/login/login-webflow.xml:105]: Add `OpenScienceFrameworkLoginHandler` with action `beforeLogin` which transits to `osfLogin` or `institutionLogin`. In the future, there will be another `osfCampaignLogin` to handle campaigns except institution in general.

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/casInstitutionLoginView.jsp (hunk 13 above)
Comment (cslzchen) [reply cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/casInstitutionLoginView.jsp:47]: For a `(KEY, VALUE)` pair in `<form:select items="${MAP}" />`, the following dropdown is generated: `<option value="KEY">VALUE</option>`, where `KEY` is the hidden login url and `VALUE` is the display name. Need to build the `MAP` with `target` in `KEY`.

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/bottom.jsp (hunk 15 above)
Comment (cslzchen) [reply cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/bottom.jsp:33]: Already changed to CAS institution login. I will leave as it is now.

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/top.jsp (hunk 18 above)
Comment (cslzchen) [reply cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/top.jsp:71]: Agreed. Leave it as it is now. Add this change for https://github.com/CenterForOpenScience/cas-overlay/pull/24.

**Comment by cslzchen** [comment]
@icereval ready to go :)

**Pull request merged**